### ISSUE TITLE
generate token 

### ISSUE THREAD

**Issue by d33pak-s1ngh** [tananaev/traccar#2763]
hi @tananaev  i have upgraded to 3.9 traccar where i can see token field i want to generate token from my site so whenever user is login a token will generate and i can use that token for accessing alert from javascript because right now to access socket data i am passing user id and password from javascript which is not good for me.
so is there any api call to get token from server for logged in user??

**Comment by tananaev** [comment]
Token is a part of user model, so you should get it with login.

**Comment by d33pak-s1ngh** [comment]
i am getting  null into password and token field when i use session api call to authenticate user

`` $res = $client->request('POST', API_PATH . "session", array(
                        'form_params' => array(
                            'email' => $userid,
                            'password' => $password
                        ),
                        'headers' => array('Content-Type' => 'application/x-www-form-urlencoded', 'Accept' => 'application/json')
                    ));``

**Comment by d33pak-s1ngh** [comment]
user get logged in successfully but token and password is null am i doing somethig wrong??

**Comment by d33pak-s1ngh** [comment]
i think i got it first i have to create token then save it and then try to access.

**Comment by tananaev** [comment]
Yes, you have have to generate a token for the user as an admin first.

**Comment by d33pak-s1ngh** [comment]
hi @tananaev  i have generated token for the user now i am trying to access socket with that but it is giving following error
url: `http://104.131.22.103:8082/api/session?token=%27UL5PeVO3BIk9wObkgN7hfMv6jqQBeh8d`
error: `NullPointerException (... < PermissionsManager:266 < SessionResource:74 < ...)`

**Comment by d33pak-s1ngh** [comment]
got it working thanx problem was with qoutes.

**Comment by iiitmahesh** [comment]
@d33pak-s1ngh  how to get token??
can you please share piece of code **OR** steps to get token.

**Comment by d33pak-s1ngh** [comment]
@iiitmahesh  you have to call a session api which will return user detail in json with token.
i am using php in which i am making api call's using guzzle and following is request i am making.

`
$res = $client->request('POST', API_PATH . "session", array( 'form_params' => array( 'email' => $userid, 'password' => $password ), 'headers' => array('Content-Type' => 'application/x-www-form-urlencoded', 'Accept' => 'application/json') ));

`
in above code i am making a post request to a session api with userid and password used for authentication which will return json in following format

 	
User {
id:
integer
name:
string
email:
string
readonly:
boolean
admin:
boolean
map:
string
distanceUnit:
string
speedUnit:
string
latitude:
number
longitude:
number
zoom:
integer
password:
string
twelveHourFormat:
boolean
coordinateFormat:
string
disabled:
boolean
expirationTime:
string
deviceLimit:
integer
token:
string
attributes:
<any>
} `

**Comment by iiitmahesh** [comment]
@d33pak-s1ngh  thanks your response.
ref: 
[https://www.traccar.org/api-reference/#paths_session_POST:](https://www.traccar.org/api-reference/#paths_session_POST)

But I am getting token as null, is there any config to enable token and set validity of token and all
```
var settings = {
  "async": true,
  "crossDomain": true,
  "url": "http://serverurl/api/session",
  "method": "POST",
  "headers": {
    "authorization": "Basic XXXXXXXXXXXX",
    "content-type": "application/x-www-form-urlencoded",
    "cache-control": "no-cache",
  },
  "data": {
    "email": "user@gps.in",
    "password": "1234"
  }
}

$.ajax(settings).done(function (response) {
  console.log(response);
});
```
**Response** : 
```
{
  "id": 2,
  "attributes": {},
  "name": "user",
  "email": "user@gps.in",
  "readonly": false,
  "admin": false,
  "map": null,
  "distanceUnit": null,
  "speedUnit": null,
  "latitude": 0,
  "longitude": 0,
  "zoom": 0,
  "twelveHourFormat": false,
  "coordinateFormat": null,
  "disabled": false,
  "expirationTime": null,
  "deviceLimit": -1,
  "userLimit": 0,
  "deviceReadonly": false,
  "token": null,
  "password": null
}
```

**Comment by tananaev** [comment]
You have to generate a token for user. Traccar doesn't do it for you.

**Comment by d33pak-s1ngh** [comment]
you have to first generate token
go to your account and generate token check following image
![token](https://cloud.githubusercontent.com/assets/23626544/24185762/ec8ad0de-0efa-11e7-8a5a-06d295770047.png)

**Comment by iiitmahesh** [comment]
@tananaev  is there any default validity for token.
how to use token in **positions API call**

**Comment by tananaev** [comment]
At the moment you can only use it to create a session.

**Comment by kgrosvenor** [comment]
How come API doesn't automatically create the token and why is it that the client does this ?

**Comment by GuljahanG** [comment]
hi, how to get user token for call api?

**Comment by tananaev** [comment]
Use the forum for questions.
